File: F#/Mathematics.fs
module Synapses.Mathematics

open FSharpx.Collections

let dotProduct
        (xs: LazyList<float>)
        (ys: LazyList<float>)
        : float =
        LazyList.map2 (fun x y -> x * y) xs ys
        |> LazyList.fold (fun acc z -> acc + z) 0.0

let euclideanDistance
        (x: LazyList<float>)
        (y: LazyList<float>)
        : float =
        LazyList.map2
                (fun xi yi ->
                        (xi - yi) * (xi - yi)
                )
                x
                y
        |> LazyList.fold (fun acc z -> acc + z) 0.0
        |> sqrt

let rootMeanSquareError
        (yHatsWithYs: LazyList<LazyList<float> * LazyList<float>>)
        : float =
        let (n, sum) =
                LazyList.map
                        (fun (yHat, y) ->
                                let d = euclideanDistance yHat y
                                d * d
                        )
                        yHatsWithYs
                |> LazyList.fold
                        (fun (accI, accS) sd -> (accI + 1, accS + sd))
                        (0, 0.0)
        sqrt (sum / float n)

File: Scala/src/main/scala/synapses/model/Mathematics.scala
package synapses.model

object Mathematics {

  def dotProduct(xs: LazyList[Double], ys: LazyList[Double]): Double =
    xs.zip(ys)
      .map { case (x, y) => x * y }
      .sum

  def euclideanDistance(xs: LazyList[Double], ys: LazyList[Double]): Double = {
    val s = xs
      .zip(ys)
      .map { case (x, y) => (x - y) * (x - y) }
      .sum
    Math.sqrt(s)
  }

  def rootMeanSquareError(yHatsWithYs: LazyList[(LazyList[Double], LazyList[Double])])
  : Double = {
    val (n, s) = yHatsWithYs
      .map { case (y_hat, y) =>
        val d = euclideanDistance(y_hat, y)
        d * d
      }
      .foldLeft((0, 0.0)) { case ((accN, accS), sd) =>
        (accN + 1, accS + sd)
      }
    Math.sqrt(s / n)
  }

}
